File: data/algorithms/Algorithms.cpp
// ============================================================
//  C++ Algorithms — Complete Implementations
// ============================================================

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <algorithm>
#include <climits>
#include <string>

// ─────────────────────────────────────────────
// SECTION 1: Searching
// ─────────────────────────────────────────────

int linearSearch(const std::vector<int>& arr, int target) {
    for (int i = 0; i < (int)arr.size(); i++)
        if (arr[i] == target) return i;
    return -1;
}

int binarySearch(const std::vector<int>& arr, int target) {
    int left = 0, right = (int)arr.size() - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) return mid;
        else if (arr[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

// ─────────────────────────────────────────────
// SECTION 2: Sorting
// ─────────────────────────────────────────────

void bubbleSort(std::vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) { std::swap(arr[j], arr[j + 1]); swapped = true; }
        }
        if (!swapped) break;
    }
}

void merge(std::vector<int>& arr, int l, int m, int r) {
    std::vector<int> L(arr.begin() + l, arr.begin() + m + 1);
    std::vector<int> R(arr.begin() + m + 1, arr.begin() + r + 1);
    int i = 0, j = 0, k = l;
    while (i < (int)L.size() && j < (int)R.size())
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    while (i < (int)L.size()) arr[k++] = L[i++];
    while (j < (int)R.size()) arr[k++] = R[j++];
}

void mergeSort(std::vector<int>& arr, int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}

int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high], i = low - 1;
    for (int j = low; j < high; j++)
        if (arr[j] <= pivot) std::swap(arr[++i], arr[j]);
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

// ─────────────────────────────────────────────
// SECTION 3: Graph Algorithms
// ─────────────────────────────────────────────

void bfs(const std::vector<std::vector<int>>& adj, int start, int n) {
    std::vector<bool> visited(n, false);
    std::queue<int> q;
    visited[start] = true;
    q.push(start);
    std::cout << "BFS: ";
    while (!q.empty()) {
        int v = q.front(); q.pop();
        std::cout << v << " ";
        for (int u : adj[v])
            if (!visited[u]) { visited[u] = true; q.push(u); }
    }
    std::cout << "\n";
}

void dfs(const std::vector<std::vector<int>>& adj, int v, std::vector<bool>& visited) {
    visited[v] = true;
    std::cout << v << " ";
    for (int u : adj[v])
        if (!visited[u]) dfs(adj, u, visited);
}

std::vector<int> dijkstra(const std::vector<std::vector<std::pair<int,int>>>& adj, int src, int n) {
    std::vector<int> dist(n, INT_MAX);
    std::priority_queue<std::pair<int,int>, std::vector<std::pair<int,int>>, std::greater<>> pq;
    dist[src] = 0;
    pq.push({0, src});
    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;
        for (auto [w, v] : adj[u]) {
            if (dist[u] + w < dist[v]) { dist[v] = dist[u] + w; pq.push({dist[v], v}); }
        }
    }
    return dist;
}

// ─────────────────────────────────────────────
// SECTION 4: Dynamic Programming
// ─────────────────────────────────────────────

std::unordered_map<int, long long> memo;
long long fib(int n) {
    if (n <= 1) return n;
    if (memo.count(n)) return memo[n];
    return memo[n] = fib(n - 1) + fib(n - 2);
}

int knapsack(int W, const std::vector<int>& wt, const std::vector<int>& val, int n) {
    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(W + 1, 0));
    for (int i = 1; i <= n; i++)
        for (int w = 0; w <= W; w++) {
            dp[i][w] = dp[i-1][w];
            if (wt[i-1] <= w)
                dp[i][w] = std::max(dp[i][w], dp[i-1][w - wt[i-1]] + val[i-1]);
        }
    return dp[n][W];
}

int lcs(const std::string& a, const std::string& b) {
    int m = a.size(), n = b.size();
    std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));
    for (int i = 1; i <= m; i++)
        for (int j = 1; j <= n; j++)
            dp[i][j] = (a[i-1] == b[j-1]) ? dp[i-1][j-1] + 1
                                            : std::max(dp[i-1][j], dp[i][j-1]);
    return dp[m][n];
}

// ─────────────────────────────────────────────
// MAIN
// ─────────────────────────────────────────────

void printVec(const std::vector<int>& v) {
    for (int x : v) std::cout << x << " ";
    std::cout << "\n";
}

int main() {
    std::cout << "===== C++ Algorithms Demo =====\n\n";

    // --- Searching ---
    std::cout << "--- Searching ---\n";
    std::vector<int> arr = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};
    std::cout << "Array: "; printVec(arr);
    std::cout << "Linear search for 23: index " << linearSearch(arr, 23) << "\n";
    std::cout << "Binary search for 56: index " << binarySearch(arr, 56) << "\n";
    std::cout << "Binary search for 99: index " << binarySearch(arr, 99) << "\n";

    // --- Sorting ---
    std::cout << "\n--- Sorting ---\n";
    std::vector<int> data = {64, 34, 25, 12, 22, 11, 90};
    std::cout << "Original: "; printVec(data);

    std::vector<int> b1 = data;
    bubbleSort(b1); std::cout << "Bubble Sort: "; printVec(b1);

    std::vector<int> b2 = data;
    mergeSort(b2, 0, b2.size() - 1); std::cout << "Merge Sort:  "; printVec(b2);

    std::vector<int> b3 = data;
    quickSort(b3, 0, b3.size() - 1); std::cout << "Quick Sort:  "; printVec(b3);

    // --- Graph ---
    std::cout << "\n--- Graph BFS & DFS ---\n";
    int n = 6;
    std::vector<std::vector<int>> adj(n);
    auto addEdge = [&](int u, int v){ adj[u].push_back(v); adj[v].push_back(u); };
    addEdge(0,1); addEdge(0,2); addEdge(1,3); addEdge(2,4); addEdge(3,5);
    bfs(adj, 0, n);
    std::vector<bool> vis(n, false);
    std::cout << "DFS: ";
    dfs(adj, 0, vis);
    std::cout << "\n";

    // --- Dijkstra ---
    std::cout << "\n--- Dijkstra Shortest Path ---\n";
    int nodes = 5;
    std::vector<std::vector<std::pair<int,int>>> wadjc(nodes);
    auto addW = [&](int u, int v, int w){ wadjc[u].push_back({w,v}); wadjc[v].push_back({w,u}); };
    addW(0,1,10); addW(0,2,3); addW(1,3,2); addW(2,1,4); addW(2,3,8); addW(2,4,2); addW(3,4,5);
    auto dist = dijkstra(wadjc, 0, nodes);
    std::cout << "Distances from node 0: ";
    for (int d : dist) std::cout << d << " "; std::cout << "\n";

    // --- Dynamic Programming ---
    std::cout << "\n--- Dynamic Programming ---\n";
    for (int i = 0; i <= 10; i++) std::cout << "fib(" << i << ")=" << fib(i) << "  ";
    std::cout << "\n";

    std::vector<int> wt = {1, 3, 4, 5};
    std::vector<int> val = {1, 4, 5, 7};
    std::cout << "Knapsack (W=7): max value = " << knapsack(7, wt, val, 4) << "\n";

    std::cout << "LCS(\"ABCBDAB\", \"BDCAB\") = " << lcs("ABCBDAB", "BDCAB") << "\n";

    std::cout << "\n✅ All algorithm demos complete!\n";
    return 0;
}

File: data/algorithms/Algorithms.rs
// ============================================================
//  Rust Algorithms — Complete Examples
// ============================================================

use std::collections::{HashMap, HashSet, VecDeque};

// ─────────────────────────────────────────────
// SECTION 1: Searching
// ─────────────────────────────────────────────

fn binary_search<T: Ord>(arr: &[T], target: &T) -> Option<usize> {
    let (mut left, mut right) = (0, arr.len());
    while left < right {
        let mid = left + (right - left) / 2;
        match arr[mid].cmp(target) {
            std::cmp::Ordering::Equal   => return Some(mid),
            std::cmp::Ordering::Less    => left  = mid + 1,
            std::cmp::Ordering::Greater => right = mid,
        }
    }
    None
}

// ─────────────────────────────────────────────
// SECTION 2: Sorting
// ─────────────────────────────────────────────

fn merge_sort(arr: &mut Vec<i32>) {
    let len = arr.len();
    if len <= 1 { return; }
    let mid   = len / 2;
    let mut l = arr[..mid].to_vec();
    let mut r = arr[mid..].to_vec();
    merge_sort(&mut l);
    merge_sort(&mut r);
    let (mut i, mut j, mut k) = (0, 0, 0);
    while i < l.len() && j < r.len() {
        if l[i] <= r[j] { arr[k] = l[i]; i += 1; } else { arr[k] = r[j]; j += 1; }
        k += 1;
    }
    while i < l.len() { arr[k] = l[i]; i += 1; k += 1; }
    while j < r.len() { arr[k] = r[j]; j += 1; k += 1; }
}

fn partition(arr: &mut [i32]) -> usize {
    let pivot = arr[arr.len() - 1];
    let mut i = 0;
    for j in 0..arr.len() - 1 {
        if arr[j] <= pivot { arr.swap(i, j); i += 1; }
    }
    arr.swap(i, arr.len() - 1);
    i
}

fn quick_sort(arr: &mut [i32]) {
    if arr.len() <= 1 { return; }
    let p = partition(arr);
    quick_sort(&mut arr[..p]);
    quick_sort(&mut arr[p + 1..]);
}

// ─────────────────────────────────────────────
// SECTION 3: Graph Algorithms
// ─────────────────────────────────────────────

fn bfs<'a>(graph: &'a HashMap<&str, Vec<&'a str>>, start: &'a str) -> Vec<String> {
    let mut visited = HashSet::new();
    let mut queue   = VecDeque::new();
    let mut order   = Vec::new();
    visited.insert(start);
    queue.push_back(start);
    while let Some(node) = queue.pop_front() {
        order.push(node.to_string());
        if let Some(neighbors) = graph.get(node) {
            for &n in neighbors { if visited.insert(n) { queue.push_back(n); } }
        }
    }
    order
}

fn dfs_helper<'a>(graph: &HashMap<&'a str, Vec<&'a str>>, node: &'a str,
                   visited: &mut HashSet<&'a str>, order: &mut Vec<String>) {
    if !visited.insert(node) { return; }
    order.push(node.to_string());
    if let Some(neighbors) = graph.get(node) {
        for &n in neighbors { dfs_helper(graph, n, visited, order); }
    }
}

fn dfs(graph: &HashMap<&str, Vec<&str>>, start: &str) -> Vec<String> {
    let mut visited = HashSet::new();
    let mut order   = Vec::new();
    dfs_helper(graph, start, &mut visited, &mut order);
    order
}

// ─────────────────────────────────────────────
// SECTION 4: Dynamic Programming
// ─────────────────────────────────────────────

fn fib(n: u64, memo: &mut HashMap<u64, u64>) -> u64 {
    if n <= 1 { return n; }
    if let Some(&v) = memo.get(&n) { return v; }
    let result = fib(n - 1, memo) + fib(n - 2, memo);
    memo.insert(n, result);
    result
}

fn lcs(a: &str, b: &str) -> usize {
    let (a, b): (Vec<char>, Vec<char>) = (a.chars().collect(), b.chars().collect());
    let (m, n) = (a.len(), b.len());
    let mut dp = vec![vec![0usize; n + 1]; m + 1];
    for i in 1..=m {
        for j in 1..=n {
            dp[i][j] = if a[i-1] == b[j-1] { dp[i-1][j-1] + 1 }
                       else { dp[i-1][j].max(dp[i][j-1]) };
        }
    }
    dp[m][n]
}

fn coin_change(coins: &[u32], amount: u32) -> Option<u32> {
    let mut dp = vec![u32::MAX; (amount + 1) as usize];
    dp[0] = 0;
    for i in 1..=(amount as usize) {
        for &c in coins {
            if (c as usize) <= i && dp[i - c as usize] != u32::MAX {
                dp[i] = dp[i].min(dp[i - c as usize] + 1);
            }
        }
    }
    if dp[amount as usize] == u32::MAX { None } else { Some(dp[amount as usize]) }
}

// ─────────────────────────────────────────────
// SECTION 5: Iterator-based Algorithms
// ─────────────────────────────────────────────

fn iterator_demos() {
    let data = vec![3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];

    let sum_squares: i32 = data.iter().filter(|&&x| x % 2 == 0).map(|&x| x * x).sum();
    println!("Sum of squares of evens: {}", sum_squares);

    let max = data.iter().copied().max().unwrap_or(0);
    let min = data.iter().copied().min().unwrap_or(0);
    println!("Max: {}, Min: {}", max, min);

    let mut freq: HashMap<i32, usize> = HashMap::new();
    data.iter().for_each(|&x| *freq.entry(x).or_insert(0) += 1);
    let mut freq_vec: Vec<_> = freq.iter().collect();
    freq_vec.sort_by(|a, b| b.1.cmp(a.1));
    println!("Most frequent: {} (appears {} times)", freq_vec[0].0, freq_vec[0].1);

    let running_max: Vec<i32> = data.iter()
        .scan(i32::MIN, |m, &x| { *m = (*m).max(x); Some(*m) })
        .collect();
    println!("Running max: {:?}", running_max);
}

// ─────────────────────────────────────────────
// MAIN
// ─────────────────────────────────────────────

fn main() {
    println!("===== Rust Algorithms Demo =====\n");

    // Searching
    println!("--- Binary Search ---");
    let sorted = vec![2, 5, 8, 12, 16, 23, 38, 56, 72, 91];
    println!("Search for 23: {:?}", binary_search(&sorted, &23));
    println!("Search for 99: {:?}", binary_search(&sorted, &99));
    println!("std binary_search for 56: {:?}", sorted.binary_search(&56));

    // Sorting
    println!("\n--- Sorting ---");
    let data = vec![64, 34, 25, 12, 22, 11, 90];
    println!("Original:   {:?}", data);

    let mut ms = data.clone(); merge_sort(&mut ms);
    println!("Merge Sort: {:?}", ms);

    let mut qs = data.clone(); quick_sort(&mut qs);
    println!("Quick Sort: {:?}", qs);

    let mut ss = data.clone(); ss.sort_unstable();
    println!("sort_unstable: {:?}", ss);

    // Graph
    println!("\n--- Graph BFS & DFS ---");
    let mut graph: HashMap<&str, Vec<&str>> = HashMap::new();
    graph.insert("A", vec!["B", "C"]);
    graph.insert("B", vec!["D", "E"]);
    graph.insert("C", vec!["F"]);
    graph.insert("D", vec![]);
    graph.insert("E", vec![]);
    graph.insert("F", vec![]);
    println!("BFS from A: {:?}", bfs(&graph, "A"));
    println!("DFS from A: {:?}", dfs(&graph, "A"));

    // Dynamic Programming
    println!("\n--- Dynamic Programming ---");
    let mut memo = HashMap::new();
    let fibs: Vec<u64> = (0..=10).map(|i| fib(i, &mut memo)).collect();
    println!("Fibonacci(0..10): {:?}", fibs);
    println!("LCS(\"ABCBDAB\",\"BDCAB\"): {}", lcs("ABCBDAB", "BDCAB"));
    println!("Coin change(11, [1,5,6,9]): {:?}", coin_change(&[1,5,6,9], 11));

    // Iterator algorithms
    println!("\n--- Iterator Algorithms ---");
    iterator_demos();

    println!("\n✅ All algorithm demos complete!");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test] fn test_binary_search_found()     { assert_eq!(binary_search(&[1,3,5,7,9], &5), Some(2)); }
    #[test] fn test_binary_search_not_found() { assert_eq!(binary_search(&[1,3,5], &4), None); }
    #[test] fn test_merge_sort() { let mut v = vec![3,1,4,1,5]; merge_sort(&mut v); assert_eq!(v, vec![1,1,3,4,5]); }
    #[test] fn test_quick_sort() { let mut v = vec![9,3,6,1,8]; quick_sort(&mut v); assert_eq!(v, vec![1,3,6,8,9]); }
    #[test] fn test_lcs() { assert_eq!(lcs("ABCBDAB","BDCAB"), 4); }
    #[test] fn test_coin_change() { assert_eq!(coin_change(&[1,5,6,9], 11), Some(2)); }
    #[test] fn test_coin_impossible() { assert_eq!(coin_change(&[2], 3), None); }
}
